### ISSUE TITLE
iroh: provide a way to supply custom TLS connection code for relay

### ISSUE THREAD

**Issue by link2xt** [n0-computer/iroh#2901]
In Delta Chat we have custom code to setup TLS connections that supports built-in DNS cache and various types of proxies: SOCKS5, HTTP(S) proxies and Shadowsocks. This code is used for all IMAP, SMTP and HTTPS connections except for the connection to iroh relay. Because of this relay connection is less reliable, e.g. it will not work if DNS is down and it cannot work over Shadowsocks which is not supported by iroh-net.

If's fine for STUN and P2P QUIC connection to be outside of control for iroh-net library user, but for relay we need some way to provide custom code to establish TLS connections that can reuse existing DNS caches, certificate checking code and so on. iroh can then run HTTP stream on top of generic stream.

**Comment by flub** [comment]
This would presumably also need to be enabled for the various HTTPS probes in netcheck?

- captive portal check
- https probe (ping) check

**Comment by dignifiedquire** [comment]
also what about general DNS resolution of the relay urls, and in the discovery service?

**Comment by matheus23** [comment]
> In Delta Chat we have custom code to setup TLS connections that supports built-in DNS cache and various types of proxies: SOCKS5, HTTP(S) proxies and Shadowsocks. This code is used for all IMAP, SMTP and HTTPS connections except for the connection to iroh relay.

I assume the reason you went for custom TLS setup code in favor of, say, using reqwest (given it supports SOCKS5, HTTP(s) proxies and DNS resolution customization), is that you want to also use this for IMAP and SMPT, right?

I'm thinking through what it'd mean if we double-down on reqwest in iroh-net, so we get SOCKS5, HTTP(s) proxies, etc. all out of the box for *all* requests we make in iroh-net, and have them be configurable in *one* way.

**Comment by matheus23** [comment]
> Does iroh resolve relay URLs without the intention to establish a connection to the relay?

Yes. At the time that dig wrote this, we used to probe the relay IPs using ICMP and STUN. We don't do that now anymore, but instead we QAD-probe them ("QUIC address discovery" - i.e. uses UDP). This gives us the closest relay plus our public IP addresses that the relays see. (If we can't QAD, we use normal HTTPS to probe the relays and find the closest one.)

> I don't know what captive portal check is for. If it is about public Wi-Fi captive portal, then why does iroh need to know about it? User has to open it in a browser anyway.

It informs iroh about whether connectivity to relays should be expected or not (I don't think it's terribly important at the moment).

---

I think we've so far not been very successful at finding an abstraction that works for all our use-cases. A single function/`trait Dialer` for establishing a TLS-ed connection doesn't quite work, since that also implies TCP, or that we open a socket at all (we might only want to do DNS TXT lookups!).

> What I want is avoiding situation where some server is reachable over HTTPS, SMTP and IMAP, but connection to relay fails because DNS is down, timeouts are lower, IP addresses are tried in different order or not at all, TLS certificate is considered invalid while we have it in our certificate storage etc.

FWIW, we've made DNS fully pluggable via a trait in #3473, and made TLS roots configurable in #3973.
Perhaps that sufficiently improves the situation already?

**Comment by link2xt** [comment]
For TLS configuring TLS roots is not sufficient because we have custom logic to keep connections to known servers working if certificate expires, e.g. during internet shutdowns: https://github.com/chatmail/core/pull/8086
